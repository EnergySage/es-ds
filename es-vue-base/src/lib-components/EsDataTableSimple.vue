<template>
    <b-table-simple
        borderless
        responsive
        :striped="striped"
        :table-class="tableClass">
        <b-tbody>
            <template v-if="isSingleCol">
                <b-tr
                    v-for="(item, index) in computedItems"
                    :key="index">
                    <b-td>
                        {{ item }}
                    </b-td>
                </b-tr>
            </template>
            <template v-else>
                <b-tr
                    v-for="item in computedItems"
                    :key="item[0]">
                    <b-td class="col-sm-7">
                        {{ item[0] }}
                    </b-td>
                    <b-td class="col-sm-4">
                        <b>{{ item[1] }}</b>
                    </b-td>
                </b-tr>
            </template>
        </b-tbody>
    </b-table-simple>
</template>

<script lang="js">
import {
    BTableSimple, BTbody, BTr, BTd,
} from 'bootstrap-vue';

export default {
    name: 'EsDataTableSimple',
    components: {
        BTableSimple, BTbody, BTr, BTd,
    },
    props: {
        fields: {
            type: Array,
            default: null,
        },
        items: {
            type: Array,
            required: true,
        },
        tableClass: {
            type: [Array, String],
            default: null,
        },
        striped: {
            type: Boolean,
            default: false,
        },
    },
    computed: {
        isSingleCol() {
            return !this.fields;
        },
        computedItems() {
            if (!this.fields) {
                return this.items;
            }

            if (this.fields.length !== this.items.length) {
                throw new Error('The number of fields and items must be the same.');
            }

            return this.fields.map((field, index) => [field, this.items[index]]);
        },
    },
};
</script>
