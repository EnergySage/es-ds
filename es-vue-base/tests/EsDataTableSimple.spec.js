import { mount } from '@vue/test-utils';
import EsDataTableSimple from '@/src/lib-components/EsDataTableSimple.vue';
import jestVue from '@/tests/jest.vue.config';

describe('EsDataTableSimple', () => {
    // Basic test; should exist for most components
    test('<EsDataTableSimple items="[\'It1\', \'It2\']" />', async () => {
        const wrapper = mount(EsDataTableSimple, {
            ...jestVue,
            propsData: {
                items: ['It1', 'It2'],
            },
        });
        const a11y = await axe(wrapper.element);

        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        expect(a11y).toHaveNoViolations();
    });

    // Test default prop values
    test('<EsDataTableSimple items="[\'It1\', \'It2\']" /> default prop values', async () => {
        const wrapper = mount(EsDataTableSimple, {
            ...jestVue,
            propsData: {
                items: ['It1', 'It2'],
            },
        });
        expect(wrapper.vm).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();

        expect(wrapper.props('fields')).toBeNull();
        expect(wrapper.props('tableClass')).toBeNull();
        expect(wrapper.props('striped')).toBeFalsy();
    });

    // Test props are there
    test('<EsDataTableSimple props />', async () => {
        const wrapper = mount(EsDataTableSimple, {
            ...jestVue,
            propsData: {
                fields: ['Fld1', 'Fld2'],
                items: ['It1', 'It2'],
                tableClass: 'cls-for-table',
                striped: true,
            },
        });

        expect(wrapper.props('fields')).toStrictEqual(['Fld1', 'Fld2']);
        expect(wrapper.props('items')).toStrictEqual(['It1', 'It2']);
        expect(wrapper.props('tableClass')).toBe('cls-for-table');
        expect(wrapper.props('striped')).toBeTruthy();

        const table = wrapper.findComponent(EsDataTableSimple);
        expect(table.text()).toContain('Fld1');
        expect(table.text()).toContain('It1');
        expect(table.find('table').classes()).toContain('table-striped');
    });

    test('<EsDataTableSimple fields="[...]" items="[...]" /> error fields and items length not matching', async () => {
        expect(() => {
            mount(EsDataTableSimple, {
                ...jestVue,
                propsData: {
                    fields: ['Fld1', 'Fld2'],
                    items: ['It1'],
                },
            });
        }).toThrow('The number of fields and items must be the same.');

        expect(() => {
            mount(EsDataTableSimple, {
                ...jestVue,
                propsData: {
                    fields: ['Fld1'],
                    items: ['It1', 'It2'],
                },
            });
        }).toThrow('The number of fields and items must be the same.');
    });

    test('<EsDataTableSimple /> error neither items prop nor slot content is defined', async () => {
        expect(() => {
            mount(EsDataTableSimple, {
                ...jestVue,
            });
        }).toThrow('Neither items prop nor slot content is defined.');
    });

    // Test slot content is there
    test('<EsDataTableSimple>Table Content</EsDataTableSimple>', async () => {
        const wrapper = mount(EsDataTableSimple, {
            ...jestVue,
            slots: {
                default: 'Table Content',
            },
        });

        expect(wrapper.props('items')).toStrictEqual([]);
        expect(wrapper.vm.$slots.default[0].text).toBe('Table Content');
        expect(wrapper.html()).toMatchSnapshot();
    });
});
