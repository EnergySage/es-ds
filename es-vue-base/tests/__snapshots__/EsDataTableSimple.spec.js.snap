// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EsDataTableSimple <EsDataTableSimple items="['It1', 'It2']" /> 1`] = `
"<div class="table-responsive">
  <table role="table" class="table b-table table-borderless" id="__BVID__4">
    <tbody role="rowgroup">
      <tr role="row" class="">
        <td role="cell" class="">
          It1
        </td>
      </tr>
      <tr role="row" class="">
        <td role="cell" class="">
          It2
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`EsDataTableSimple <EsDataTableSimple items="['It1', 'It2']" /> default prop values 1`] = `
"<div class="table-responsive">
  <table role="table" class="table b-table table-borderless">
    <tbody role="rowgroup">
      <tr role="row" class="">
        <td role="cell" class="">
          It1
        </td>
      </tr>
      <tr role="row" class="">
        <td role="cell" class="">
          It2
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`EsDataTableSimple <EsDataTableSimple>Table Content</EsDataTableSimple> 1`] = `
"<div class="table-responsive">
  <table role="table" class="table b-table table-borderless">Table Content</table>
</div>"
`;
